src/V3LinkInc.cpp

```diff
@@ -0,0 +1,239 @@
+        IM_WHILE_PRECOND  // Pointing to for loop, add to body end
+    };
+
+    // STATE
+    int m_modIncrementsNum;  // Var name counter
+    InsertMode m_insMode;  // How to insert
+    AstNode* m_insStmtp;  // Where to insert statement
+    bool m_unsupportedHere;  // Used to detect where it's not supported yet
+
+private:
+    void insertBeforeStmt(AstNode* nodep, AstNode* newp) {
+        // Return node that must be visited, if any
+        // See also AstNode::addBeforeStmt; this predates that function
+        if (debug() >= 9) newp->dumpTree(cout, "-newstmt:");
+        UASSERT_OBJ(m_insStmtp, nodep, "Function not underneath a statement");
+        if (m_insMode == IM_BEFORE) {
+            // Add the whole thing before insertAt
+            if (debug() >= 9) newp->dumpTree(cout, "-newfunc:");
+            m_insStmtp->addHereThisAsNext(newp);
+        } else if (m_insMode == IM_AFTER) {
+            m_insStmtp->addNextHere(newp);
+        } else if (m_insMode == IM_WHILE_PRECOND) {
+            AstWhile* whilep = VN_CAST(m_insStmtp, While);
+            UASSERT_OBJ(whilep, nodep, "Insert should be under WHILE");
+            whilep->addPrecondsp(newp);
+        } else {
+            nodep->v3fatalSrc("Unknown InsertMode");
+        }
+        m_insMode = IM_AFTER;
+        m_insStmtp = newp;
+    }
+
+    // VISITORS
+    virtual void visit(AstWhile* nodep) VL_OVERRIDE {
+        // Special, as statements need to be put in different places
+        // Preconditions insert first just before themselves (the normal
+        // rule for other statement types)
+        m_insStmtp = NULL;  // First thing should be new statement
+        iterateAndNextNull(nodep->precondsp());
+        // Conditions insert first at end of precondsp.
+        m_insMode = IM_WHILE_PRECOND;
+        m_insStmtp = nodep;
+        iterateAndNextNull(nodep->condp());
+        // Body insert just before themselves
+        m_insStmtp = NULL;  // First thing should be new statement
+        iterateAndNextNull(nodep->bodysp());
+        iterateAndNextNull(nodep->incsp());
+        // Done the loop
+        m_insStmtp = NULL;  // Next thing should be new statement
+    }
+    virtual void visit(AstNodeFor* nodep) VL_OVERRIDE {
+        nodep->v3fatalSrc(
+            "For statements should have been converted to while statements in V3Begin.cpp");
+    }
+    virtual void visit(AstNodeStmt* nodep) VL_OVERRIDE {
+        if (!nodep->isStatement()) {
+            iterateChildren(nodep);
+            return;
+        }
+        m_insMode = IM_BEFORE;
+        m_insStmtp = nodep;
+        iterateChildren(nodep);
+        m_insStmtp = NULL;  // Next thing should be new statement
+    }
+    void unsupported_visit(AstNode* nodep) {
+        m_unsupportedHere = true;
+        iterateChildren(nodep);
+        m_unsupportedHere = false;
+    }
+    virtual void visit(AstLogAnd* nodep) VL_OVERRIDE { unsupported_visit(nodep); }
```

Comment:
Need to either do the "unsupported unless understood" scheme we discussed, or you need to audit all possible node types for those that can cause problems. (There are other short circuiting operators, see IEEE.)

Reply:
Should there be new `_bad` test cases for other unsupported node types? Or is there a way to group these into one test?

Reply:
Yes, we should test any code path that is here, so for example if you need to test && and || then both can be in the same test (make sure you get two errors out).  Sometimes we need multiple _bad's when the results are fatal or in different stages, but shouldn't be the case here.